frontend/src/pages/GraphicalSelector/GraphicalSelector.jsx

```diff
@@ -1,72 +1,132 @@
+    ).data;
+    const courseList = (
+      Object.values(structure)
+        .flatMap((specialisation) => Object.values(specialisation)
+          .filter((spec) => typeof spec === "object" && spec.courses && !spec.type.includes("rule"))
+          .flatMap((spec) => Object.keys(spec.courses)))
+    );
+    const res = await Promise.all(courseList.map((c) => axios.get(`/courses/getPathFrom/${c}`).catch((e) => e)));
+    // filter any errors from res
+    const children = res.filter((value) => value?.data?.courses).map((value) => value.data);
+    const edges = children
+      .flatMap((courseObject) => courseObject.courses
+        .filter((course) => courseList.includes(course))
+        .map((course) => ({ source: courseObject.original, target: course })));
+    if (courseList.length !== 0 && edges.length !== 0) initialiseGraph(courseList, edges);
+    setLoading(false);
   };
-  useEffect(() => {
-    const fetchCourseList = async () => {
-      const res = (
-        await axios.get(`/programs/getStructure/${programCode}/${specs.join("+")}}`)
-      ).data;
-      const courseList = (
-        Object.values(res.structure)
-          .flatMap((specialisation) => Object.values(specialisation)
-            .filter((spec) => typeof spec === "object" && spec.courses)
-            .flatMap((spec) => Object.keys(spec.courses)))
-      );
-      setCourses(courseList);
-      const res2 = await Promise.all(courseList.map((c) => axios.get(`/courses/getPathFrom/${c}`)));
-      const children = res2.map((value) => value.data);
-      const edges = children.flatMap(
-        (courseObject) => courseObject.courses.filter((course) => courseList.includes(course)).map(
-          (course) => ({ source: courseObject.original, target: course }),
-        ),
-      );
-      setCourseEdges(edges);
-    };
-    if (programCode) fetchCourseList();
-  }, [programCode, specs]);
 
   useEffect(() => {
-    if (courses.length !== 0 && courseEdges.length !== 0) loadGraph();
-  }, [courses, courseEdges]);
-  return <PageTemplate><div ref={ref} /></PageTemplate>;
+    if (programCode) setupGraph();
+  }, []);
+
+  return (
+    <PageTemplate>
+      {loading ? "This page is loading..." : (
+        <>
+          <button
+            type="button"
+            onClick={() => {
+              const nodes = graph.getNodes();
+              const edges = graph.getEdges();
+              nodes.forEach((n) => n.hide());
+              edges.forEach((e) => e.hide());
+            }}
+          >
+            Hide all nodes and edges
+          </button>
+          <button
+            type="button"
+            onClick={() => {
+              const nodes = graph.getNodes();
+              const edges = graph.getEdges();
+              nodes.forEach((n) => n.show());
+              edges.forEach((e) => e.show());
+            }}
+          >
+            Show all nodes and edges
+          </button>
+        </>
+      )}
+      <S.Wrapper ref={ref} />
```

Comment:
Wait does this fix the duplicate on reload bug? I noticed that but couldn't be fucked fixing it

Reply:
It doesn't fix the duplicate thing unfortunately. might be just an antv thing rip